core/sdl/dreampicoport.cpp

```diff
@@ -625,498 +653,843 @@ class DreamPicoPortSerialHandler
+		u8 hwDestAP = (hardware_bus << 6) | (msg.destAP & 0x3F);
+		u8 hwOriginAP = (hardware_bus << 6) | (msg.originAP & 0x3F);
+		tx.packet.push_back(hwDestAP);
+		tx.packet.push_back(hwOriginAP);
+		tx.packet.push_back(msg.size);
+		tx.packet.insert(tx.packet.end(), msg.data, msg.data + data_size);
+		const uint64_t id = dpp_api_device->send(tx);
+		return (id != 0);
 	}
 
-	++connected_dev_count;
-	connection_established = true;
-	if (!serial) {
-		serial = std::make_unique<DreamPicoPortSerialHandler>();
+    bool send(const MapleMsg& txMsg, MapleMsg& rxMsg, std::chrono::milliseconds timeout_ms) override {
+		if (!isConnected()) {
+			return false;
+		}
+
+		dpp_api::msg::tx::Maple tx;
+		tx.emu = true;
+		const u32 data_size = txMsg.getDataSize();
+		tx.packet.reserve(data_size + 4);
+		tx.packet.push_back(txMsg.command);
+		// Need to message the hardware bus instead of the software bus
+		u8 hwDestAP = (hardware_bus << 6) | (txMsg.destAP & 0x3F);
+		u8 hwOriginAP = (hardware_bus << 6) | (txMsg.originAP & 0x3F);
+		tx.packet.push_back(hwDestAP);
+		tx.packet.push_back(hwOriginAP);
+		tx.packet.push_back(txMsg.size);
+		tx.packet.insert(tx.packet.end(), txMsg.data, txMsg.data + data_size);
+		dpp_api::msg::rx::Maple rx = dpp_api_device->sendSync(tx, 100);
+		if (rx.cmd != dpp_api::msg::rx::Msg::kCmdSuccess || rx.packet.size() < 4) {
+			return false;
+		}
+		rxMsg.command = rx.packet[0];
+		rxMsg.destAP = rx.packet[1];
+		rxMsg.originAP = rx.packet[2];
+		rxMsg.size = rx.packet[3];
+		if (rx.packet.size() > 4) {
+			memcpy(rxMsg.data, &rx.packet[4], (std::min)(rx.packet.size() - 4, sizeof(rxMsg.data)));
+		}
+		return (rxMsg.getDataSize() <= (rx.packet.size() - 4));
 	}
 
-	if (serial && serial->is_open()) {
-		sendPort();
-	} else {
-		disconnect();
-		return;
+	void sendPort(std::chrono::milliseconds timeout_ms) override {
+		dpp_api::msg::tx::ChangePlayerDisplay changePlayerDisplay;
+		changePlayerDisplay.idx = hardware_bus;
+		changePlayerDisplay.toIdx = software_bus;
+		dpp_api_device->send(changePlayerDisplay, nullptr);
 	}
+};
 
-	if (!queryInterfaceVersion()) {
-		disconnect();
-		return;
+std::unordered_map<std::string, std::weak_ptr<dpp_api::DppDevice>> ApiDreamPicoPortComms::all_dpp_api_devices;
+std::mutex ApiDreamPicoPortComms::all_dpp_api_devices_mutex;
+
+class DreamPicoPortImp : public DreamPicoPort
+{
+	//! Implements communication interface to DreamPicoPort
+	std::unique_ptr<class DreamPicoPortComms> dpp_comms;
+	//! Current timeout in milliseconds
+	std::chrono::milliseconds timeout_ms;
+	//! The bus ID dictated by flycast
+	int software_bus = -1;
+	//! The bus index of the hardware connection which will differ from the software bus
+	int hardware_bus = -1;
+	//! true iff only a single devices was found when enumerating devices
+	bool is_single_device = true;
+	//! True when initial enumeration failed
+	bool is_hardware_bus_implied = true;
+    //! The queried interface version
+    double interface_version = 0.0;
```

Comment:
nit: It feels like a string, or separate major/minor fields (possibly in a separate "version" data type), would be more suitable, because a double can't exactly represent values like `1.1`, for example.